catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,355 @@
+    LiteralSuggestion literal = getMgrsSuggestion(query);
+    if (literal != null && literal.hasGeo()) {
+      LOGGER.debug("Adding the MGRS suggestion to results");
+      results.add(0, literal);
+    }
+    return results;
+  }
+
+  /**
+   * Given an input string, return the list of properly formatted MGRS coordinate strings that are
+   * ready for submission to usng4j for mathematical evaluation.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a list of valid MGRS strings ready for evaluation.
+   */
+  @VisibleForTesting
+  List<String> getMgrsCoordinateStrings(String query) {
+    return coordinateStringsFromQuery(query);
+  }
+
+  /**
+   * Build the MGRS suggestion.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a suggestion object with geometry attached to it; a {@link BoundingBox} if only one
+   *     coordinate was detected, or a collection of {@link LatLon} center points otherwise.
+   */
+  @Nullable
+  private LiteralSuggestion getMgrsSuggestion(String query) {
+    List<String> matches = coordinateStringsFromQuery(query);
+    if (matches.isEmpty()) {
+      LOGGER.debug("No valid MGRS strings were found or could be inferred");
+      return null;
+    }
+    StringBuilder nameBuilder = new StringBuilder("MGRS: ");
+    List<UsngCoordinate> uniqueMatches =
+        matches
+            .stream()
+            .distinct()
+            .map(this::parseMgrsString)
+            .filter(Objects::nonNull)
+            .peek(mgrs -> nameBuilder.append("[ ").append(mgrs.toMgrsString()).append(" ] "))
+            .collect(Collectors.toList());
+    if (uniqueMatches.size() == 1) {
+      UsngCoordinate mgrsCoord = uniqueMatches.get(0);
+      BoundingBox bb = translator.toBoundingBox(mgrsCoord);
+      return new LiteralSuggestion(LITERAL, nameBuilder.toString(), latLonFromBoundingBox(bb));
+    }
+    List<LatLon> geo =
+        uniqueMatches
+            .stream()
+            .map(translator::toLatLon)
+            .map(d -> from(d.getLat(), d.getLon()))
+            .collect(Collectors.toList());
+    return new LiteralSuggestion(LITERAL, nameBuilder.toString(), geo);
+  }
+
+  @Nullable
+  private UsngCoordinate parseMgrsString(String mgrs) {
+    try {
+      return translator.parseMgrsString(mgrs);
+    } catch (ParseException e) {
+      LOGGER.debug(format("Detected string [%s] was not valid MGRS", mgrs), e);
+      return null;
+    }
+  }
+
+  private static List<LatLon> latLonFromBoundingBox(BoundingBox bb) {
+    return ImmutableList.of(
+        from(bb.getSouth(), bb.getWest()),
+        from(bb.getNorth(), bb.getWest()),
+        from(bb.getNorth(), bb.getEast()),
+        from(bb.getSouth(), bb.getEast()));
+  }
+
+  private static List<String> coordinateStringsFromQuery(String query) {
+    final Matcher matcher = PATTERN_MGRS_ANYPART.matcher(query.concat(" "));
```

Comment:
❓ Why do we need to add a space at the end of the query?

Reply:
I'll add javadoc to expain this. Look at the regex for the MGRS `NUMERIC` component. 

There's an edge case where something like `12S VB 137 987 14Q` ... which in a forgiving model represents two valid MGRS coordinates despite the ambiguity ... results in the `14` interpreted as part of the previous coordinate and not a new one. 

Updates to the `NUMERIC` regex resolve this but now the query itself needs to be "terminated" so valid matches will be recognized without the user having to manually append a space.

Reply:
Couldn't we cut down on user error if we changed how the user inputs data? Something like this should be considered for the future

![11213-3952-007-apple-id-verification-code-l](https://user-images.githubusercontent.com/7646577/47440667-dcc26500-d76b-11e8-81b3-aa6511531656.jpg)